File: android/MultiAIIM/app/src/main/java/com/multiaicode/remoteim/LocalChatHistoryStore.java
package com.multiaicode.remoteim;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

public final class LocalChatHistoryStore {
    private final File baseDirectory;

    public LocalChatHistoryStore(File baseDirectory) {
        this.baseDirectory = baseDirectory;
    }

    public void save(ChatState state) throws IOException {
        if (!baseDirectory.exists() && !baseDirectory.mkdirs()) {
            throw new IOException("failed to create history directory");
        }

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(historyFile(state.ownerUserId())))) {
            for (RemoteIMContact contact : state.contacts()) {
                writer.write("CONTACT");
                writer.write('\t');
                writer.write(encode(contact.userId()));
                writer.write('\t');
                writer.write(encode(contact.displayName()));
                writer.newLine();
            }

            for (RemoteIMMessage message : state.messages()) {
                writeMessage(writer, message);
            }
        }
    }

    public ChatState load(String ownerUserId) throws IOException {
        ChatState state = new ChatState(ownerUserId);
        File file = historyFile(ownerUserId);
        if (!file.exists()) return state;

        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
            String line;
            while ((line = reader.readLine()) != null) {
                String[] parts = line.split("\t", -1);
                if (parts.length == 0) continue;
                if ("CONTACT".equals(parts[0]) && parts.length >= 3) {
                    state.upsertContact(new RemoteIMContact(decode(parts[1]), decode(parts[2])));
                } else if ("MESSAGE".equals(parts[0]) && parts.length >= 14) {
                    state.addRestoredMessage(readMessage(parts));
                }
            }
        }

        return state;
    }

    private File historyFile(String ownerUserId) {
        return new File(baseDirectory, safeFileName(ownerUserId) + ".tsv");
    }

    private void writeMessage(BufferedWriter writer, RemoteIMMessage message) throws IOException {
        RemoteIMImageAttachment image = message.imageAttachment();
        RemoteIMVoiceAttachment voice = message.voiceAttachment();
        writer.write("MESSAGE");
        writer.write('\t');
        writer.write(encode(message.id()));
        writer.write('\t');
        writer.write(encode(message.fromUserId()));
        writer.write('\t');
        writer.write(encode(message.toUserId()));
        writer.write('\t');
        writer.write(encode(message.text()));
        writer.write('\t');
        writer.write(message.direction().name());
        writer.write('\t');
        writer.write(message.status().name());
        writer.write('\t');
        writer.write(Long.toString(message.createdAtMillis()));
        writer.write('\t');
        writer.write(encode(image == null ? "" : image.localPath()));
        writer.write('\t');
        writer.write(Integer.toString(image == null ? 0 : image.width()));
        writer.write('\t');
        writer.write(Integer.toString(image == null ? 0 : image.height()));
        writer.write('\t');
        writer.write(Long.toString(image == null ? 0 : image.sizeBytes()));
        writer.write('\t');
        writer.write(encode(voice == null ? "" : voice.localPath()));
        writer.write('\t');
        writer.write(Integer.toString(voice == null ? 0 : voice.durationSeconds()));
        writer.newLine();
    }

    private RemoteIMMessage readMessage(String[] parts) {
        RemoteIMImageAttachment image = null;
        String imagePath = decode(parts[8]);
        if (!imagePath.isEmpty()) {
            image = new RemoteIMImageAttachment(
                imagePath,
                parseInt(parts[9]),
                parseInt(parts[10]),
                parseLong(parts[11])
            );
        }

        RemoteIMVoiceAttachment voice = null;
        String voicePath = decode(parts[12]);
        if (!voicePath.isEmpty()) {
            voice = new RemoteIMVoiceAttachment(voicePath, parseInt(parts[13]));
        }

        return new RemoteIMMessage(
            decode(parts[1]),
            decode(parts[2]),
            decode(parts[3]),
            decode(parts[4]),
            RemoteIMMessage.Direction.valueOf(parts[5]),
            RemoteIMMessage.Status.valueOf(parts[6]),
            parseLong(parts[7]),
            image,
            voice
        );
    }

    private static String encode(String value) {
        return Base64.getUrlEncoder().encodeToString(
            (value == null ? "" : value).getBytes(StandardCharsets.UTF_8)
        );
    }

    private static String decode(String value) {
        if (value == null || value.isEmpty()) return "";
        return new String(Base64.getUrlDecoder().decode(value), StandardCharsets.UTF_8);
    }

    private static int parseInt(String value) {
        try {
            return Integer.parseInt(value);
        } catch (NumberFormatException err) {
            return 0;
        }
    }

    private static long parseLong(String value) {
        try {
            return Long.parseLong(value);
        } catch (NumberFormatException err) {
            return 0;
        }
    }

    private static String safeFileName(String value) {
        return (value == null ? "" : value.trim()).replaceAll("[^A-Za-z0-9._-]", "_");
    }
}

File: desktop/qt-im/src/storage/LocalChatHistoryStore.cpp
#include "storage/LocalChatHistoryStore.h"

#include <QDir>
#include <QFile>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <utility>

namespace {
QString directionToString(RemoteIMMessageDirection direction) {
    return direction == RemoteIMMessageDirection::Outgoing ? "outgoing" : "incoming";
}

RemoteIMMessageDirection directionFromString(const QString& direction) {
    return direction == "outgoing" ? RemoteIMMessageDirection::Outgoing : RemoteIMMessageDirection::Incoming;
}

QJsonObject contactToJson(const RemoteIMContact& contact) {
    QJsonObject object;
    object["userId"] = contact.userId;
    object["displayName"] = contact.displayName;
    return object;
}

RemoteIMContact contactFromJson(const QJsonObject& object) {
    return RemoteIMContact{object["userId"].toString(), object["displayName"].toString()};
}

QJsonObject messageToJson(const RemoteIMMessage& message) {
    QJsonObject object;
    object["id"] = message.id;
    object["from"] = message.fromUserId;
    object["to"] = message.toUserId;
    object["text"] = message.text;
    object["direction"] = directionToString(message.direction);
    object["status"] = static_cast<int>(message.status);
    object["createdAtMillis"] = QString::number(message.createdAtMillis);
    object["hasImage"] = message.hasImage;
    object["imagePath"] = message.image.localPath;
    object["imageWidth"] = message.image.width;
    object["imageHeight"] = message.image.height;
    object["imageSizeBytes"] = QString::number(message.image.sizeBytes);
    object["hasVoice"] = message.hasVoice;
    object["voicePath"] = message.voice.localPath;
    object["voiceDurationSeconds"] = message.voice.durationSeconds;
    return object;
}

RemoteIMMessage messageFromJson(const QJsonObject& object) {
    RemoteIMMessage message;
    message.id = object["id"].toString();
    message.fromUserId = object["from"].toString();
    message.toUserId = object["to"].toString();
    message.text = object["text"].toString();
    message.direction = directionFromString(object["direction"].toString());
    message.status = static_cast<RemoteIMMessageStatus>(object["status"].toInt());
    message.createdAtMillis = object["createdAtMillis"].toString().toLongLong();
    message.hasImage = object["hasImage"].toBool();
    message.image = RemoteIMImageAttachment{
        object["imagePath"].toString(),
        object["imageWidth"].toInt(),
        object["imageHeight"].toInt(),
        object["imageSizeBytes"].toString().toLongLong()
    };
    message.hasVoice = object["hasVoice"].toBool();
    message.voice = RemoteIMVoiceAttachment{
        object["voicePath"].toString(),
        object["voiceDurationSeconds"].toInt()
    };
    return message;
}
}

LocalChatHistoryStore::LocalChatHistoryStore(QString rootDir) : rootDir_(std::move(rootDir)) {}

bool LocalChatHistoryStore::save(const ChatState& state) const {
    QDir().mkpath(rootDir_);
    QJsonObject root;
    root["ownerUserId"] = state.ownerUserId();
    root["selectedPeerId"] = state.selectedPeerId();

    QJsonArray contacts;
    for (const RemoteIMContact& contact : state.contacts()) {
        contacts.append(contactToJson(contact));
    }
    root["contacts"] = contacts;

    QJsonArray messages;
    for (const RemoteIMMessage& message : state.messages()) {
        messages.append(messageToJson(message));
    }
    root["messages"] = messages;

    QFile file(filePath(state.ownerUserId()));
    if (!file.open(QIODevice::WriteOnly | QIODevice::Truncate)) return false;
    file.write(QJsonDocument(root).toJson(QJsonDocument::Compact));
    return true;
}

bool LocalChatHistoryStore::load(const QString& ownerUserId, ChatState& state) const {
    QFile file(filePath(ownerUserId));
    if (!file.exists()) return true;
    if (!file.open(QIODevice::ReadOnly)) return false;

    const QJsonDocument document = QJsonDocument::fromJson(file.readAll());
    if (!document.isObject()) return false;
    const QJsonObject root = document.object();

    ChatState loaded(ownerUserId);
    for (const QJsonValue& value : root["contacts"].toArray()) {
        loaded.upsertContact(contactFromJson(value.toObject()));
    }
    const QString selectedPeerId = root["selectedPeerId"].toString();
    if (!selectedPeerId.isEmpty()) loaded.selectPeer(selectedPeerId);

    for (const QJsonValue& value : root["messages"].toArray()) {
        loaded.appendMessageForRestore(messageFromJson(value.toObject()));
    }
    state = loaded;
    return true;
}

QString LocalChatHistoryStore::filePath(const QString& ownerUserId) const {
    return QDir(rootDir_).filePath(ownerUserId.trimmed() + ".json");
}
